packages/runtime/container-runtime/src/orderedClientElection.ts

```diff
@@ -0,0 +1,263 @@
+    private eligibleCount = 0;
+    /** Count of summarizer clients. */
+    private summarizerCount = 0;
+    /** Currently elected client (within linked list). */
+    private electedClient: IEligibleClient | undefined;
+
+    constructor(quorum: Pick<IQuorum, "getMembers" | "on">) {
+        super();
+        const members = quorum.getMembers();
+        for (const [clientId, client] of members) {
+            this.addClient(clientId, client);
+        }
+
+        quorum.on("addMember", this.addClient);
+        quorum.on("removeMember", this.removeClient);
+
+        this.resetElectedClient();
+    }
+
+    private insertEligibleClient(clientId: string, sequenceNumber: number): IEligibleClient {
+        // Normal case is adding the latest client, which will bypass loop.
+        // Find where it belongs otherwise (this shouldn't happen, assert?).
+        assert(sequenceNumber > -1, "Negative client sequence number not allowed");
+        let currClient = this.youngestClient;
+        while (currClient.sequenceNumber > sequenceNumber) {
+            assert(currClient.olderClient !== undefined, "Previous client should always be defined");
+            // what to do if currClient === this.currentClient
+            currClient = currClient.olderClient;
+        }
+
+        // Now currClient is the node right before where the new client node should be.
+        const newClient: IEligibleClient = {
+            clientId,
+            sequenceNumber,
+            eligible: true,
+            olderClient: currClient,
+            youngerClient: currClient.youngerClient,
+        };
+
+        // Update prev node to point to this new node.
+        newClient.olderClient.youngerClient = newClient;
+
+        if (newClient.youngerClient === undefined) {
+            // Update linked list end pointer to youngest client.
+            this.youngestClient = newClient;
+        } else {
+            // Update next node to point back to this new node.
+            newClient.youngerClient.olderClient = newClient;
+        }
+
+        this.eligibleCount++;
+        return newClient;
+    }
+
+    private deleteEligibleClient(removeClient: Readonly<IEligibleClient>) {
+        // Update prev node to point to next node.
+        removeClient.olderClient.youngerClient = removeClient.youngerClient;
+
+        if (removeClient.youngerClient === undefined) {
+            // Update linked list end pointer to youngest client.
+            this.youngestClient = removeClient.olderClient;
+        } else {
+            // Update next node to point back to previous node.
+            removeClient.youngerClient.olderClient = removeClient.olderClient;
+        }
+
+        this.eligibleCount--;
+    }
+
+    private readonly addClient = (clientId: string, client: ISequencedClient) => {
+        const isSummarizer = client.client.details?.type === summarizerClientType;
+        const eligible = !isSummarizer && (client.client.details?.capabilities.interactive ?? true);
```

Comment:
>true [](http://example.com/codeflow?start=94&length=4)

should this default to false?

Reply:
We actually don't even need this anymore.